File: ml/scan/pipeline/lib/src/store_classifier.dart
library;

import 'dart:convert';

import 'line_signals.dart' show crc32;
import 'lines.dart';
import 'store_gazetteer.dart' show normalizeStore;

const int storeClassifierBuckets = 1 << 16;
const String storeClassifierOther = '';

int _bucket(String token) => crc32(utf8.encode(token)) % storeClassifierBuckets;

Set<int> ticketFeatures(List<PhysicalLine> lines) {
  final features = <int>{};
  for (final line in lines) {
    final words = normalizeStore(line.text)
        .split(' ')
        .where((w) => w.isNotEmpty)
        .toList();
    for (final word in words) {
      features.add(_bucket(word));
    }
    for (var index = 0; index + 1 < words.length; index++) {
      features.add(_bucket('${words[index]} ${words[index + 1]}'));
    }
  }
  return features;
}

class StoreClassifier {
  StoreClassifier({
    required this.classes,
    required this.intercepts,
    required this.weights,
  }) {
    if (classes.length != intercepts.length ||
        classes.length != weights.length) {
      throw StateError(
        'classes, biais et poids doivent avoir la même longueur',
      );
    }
  }

  factory StoreClassifier.fromJson(Map<String, dynamic> data) {
    if (data['buckets'] != storeClassifierBuckets) {
      throw StateError(
        '${data['buckets']} seaux dans le modèle, $storeClassifierBuckets attendus',
      );
    }
    return StoreClassifier(
      classes: (data['classes'] as List<dynamic>).cast<String>(),
      intercepts: [
        for (final value in data['intercepts'] as List<dynamic>)
          (value as num).toDouble(),
      ],
      weights: [
        for (final weight in data['weights'] as List<dynamic>)
          {
            for (final entry in (weight as Map<String, dynamic>).entries)
              int.parse(entry.key): (entry.value as num).toDouble(),
          },
      ],
    );
  }

  final List<String> classes;
  final List<double> intercepts;
  final List<Map<int, double>> weights;

  List<double> scores(List<PhysicalLine> lines) {
    final features = ticketFeatures(lines);
    return [
      for (var index = 0; index < classes.length; index++)
        intercepts[index] +
            features.fold<double>(
              0.0,
              (sum, f) => sum + (weights[index][f] ?? 0.0),
            ),
    ];
  }

  String? predict(List<PhysicalLine> lines) {
    final scored = scores(lines);
    var best = 0;
    for (var index = 1; index < scored.length; index++) {
      if (scored[index] > scored[best]) best = index;
    }
    final store = classes[best];
    return store == storeClassifierOther ? null : store;
  }

  Map<String, dynamic> toJson() => {
    'buckets': storeClassifierBuckets,
    'classes': classes,
    'intercepts': intercepts,
    'weights': [
      for (final weight in weights)
        {for (final key in weight.keys.toList()..sort()) '$key': weight[key]},
    ],
  };
}

File: ml/scan/research/reference/store_classifier.py
"""L'enseigne, lue sur le ticket entier.

Le tagger désigne une ligne d'enseigne. Sur dix-sept tickets du held-out, il
n'y en a aucune à désigner : le logo n'est pas lu, et le nom n'est que dans
`www.auchan.fr`, dans « l'application CARREFOUR », dans un pied de ticket. Un
tagger de lignes ne peut pas apprendre ça, quelle que soit la quantité de
données : la question n'est pas « quelle ligne » mais « quel ticket ».

Ce classifieur pose cette question-là. Un ticket est un sac de mots et de
bigrammes hachés ; l'enseigne est une classe parmi celles que le corpus
d'entraînement connaît assez, ou « autre » — et « autre » rend la main à la
ligne désignée par le tagger. Régression logistique, poids élagués pour tenir
dans un asset ; l'inférence est une somme de poids, la même en Dart
(`pipeline/lib/src/store_classifier.dart`).

Aucun seuil de confiance : la classe « autre » est apprise comme les autres.
"""

from __future__ import annotations

import json
import zlib
from itertools import pairwise

from paths import MODELS_DIR
from reference.lines import PhysicalLine
from reference.store_gazetteer import normalize

STORE_CLASSIFIER_PATH = MODELS_DIR / "store_classifier.json"
BUCKETS = 1 << 16
OTHER = ""
# Quatre décimales suffisent à l'argmax et divisent l'asset par deux.
WEIGHT_DECIMALS = 4


def _bucket(token: str) -> int:
    return zlib.crc32(token.encode()) % BUCKETS


def ticket_features(lines: list[PhysicalLine]) -> set[int]:
    """Les mots et bigrammes normalisés du ticket, hachés. Un bigramme ne
    traverse pas une ligne : « AUCHAN » en fin de ligne et « PESSAC » au
    début de la suivante ne forment pas une expression."""
    features: set[int] = set()
    for line in lines:
        words = normalize(line.text).split()
        features.update(_bucket(word) for word in words)
        features.update(
            _bucket(f"{first} {second}") for first, second in pairwise(words)
        )
    return features


class StoreClassifier:
    """`classes[i]` est la graphie rendue ; `weights[i]` les poids non nuls de
    la classe, par trait haché ; la classe `OTHER` rend `None`."""

    def __init__(
        self,
        classes: list[str],
        intercepts: list[float],
        weights: list[dict[int, float]],
    ) -> None:
        if not len(classes) == len(intercepts) == len(weights):
            raise ValueError("classes, biais et poids doivent avoir la même longueur")
        self.classes = classes
        self.intercepts = intercepts
        self.weights = weights

    def scores(self, lines: list[PhysicalLine]) -> list[float]:
        features = ticket_features(lines)
        return [
            intercept + sum(weight.get(feature, 0.0) for feature in features)
            for intercept, weight in zip(self.intercepts, self.weights)
        ]

    def predict(self, lines: list[PhysicalLine]) -> str | None:
        scores = self.scores(lines)
        best = max(range(len(scores)), key=lambda index: scores[index])
        return self.classes[best] or None

    def to_json(self) -> dict:
        return {
            "buckets": BUCKETS,
            "classes": self.classes,
            "intercepts": [round(value, WEIGHT_DECIMALS) for value in self.intercepts],
            "weights": [
                {
                    str(feature): round(value, WEIGHT_DECIMALS)
                    for feature, value in sorted(weight.items())
                }
                for weight in self.weights
            ],
        }

    @classmethod
    def from_json(cls, data: dict) -> StoreClassifier:
        if data["buckets"] != BUCKETS:
            raise ValueError(
                f"{data['buckets']} seaux dans le modèle, {BUCKETS} attendus"
            )
        return cls(
            classes=list(data["classes"]),
            intercepts=[float(value) for value in data["intercepts"]],
            weights=[
                {int(feature): float(value) for feature, value in weight.items()}
                for weight in data["weights"]
            ],
        )


_classifier: StoreClassifier | None = None


def load() -> StoreClassifier:
    global _classifier
    if _classifier is None:
        _classifier = StoreClassifier.from_json(
            json.loads(STORE_CLASSIFIER_PATH.read_text())
        )
    return _classifier
